Onboarding Improvements: New Login Epilogue Screen Design

ParaskP7 opened the pull request:
Parent #14990

This PR adds a couple of new designs to the login epilogue screen, the `normal` and `expanded` ones.
- The `normal` design is the one that will be shown when the number of sites equal to 2-3 or less.
- The `expanded` design is the one that will be shown when the number of sites end up being more than 3.

Sites | Existing | New
-----|-----|-----
2-3 or less<br/>(`normal`) | <img width="250" height="500" alt="existing_normal" src="https://user-images.githubusercontent.com/9729923/125948450-1f6172b0-8b4c-4fca-b518-b659d2ad731f.png"> | <img width="250" height="500" alt="new_normal" src="https://user-images.githubusercontent.com/9729923/126322869-39a734e9-f55c-4778-8830-b706a89c4873.png">
More than 3<br/>(`expanded`) | <img width="250" height="500" alt="existing_expanded" src="https://user-images.githubusercontent.com/9729923/125948499-fc11eeec-1fa8-41c9-a997-7d6d0fe73ffc.png"> | <img width="250" height="500" alt="new_expanded" src="https://user-images.githubusercontent.com/9729923/126322877-025659f7-60f4-49aa-8d0e-6c9379bd0067.png">

-----

To test:

#### Scenario: `normal`
1. Find an account which has lots of sites and log-in to the app.
2. Go to the `Side Chooser` screen and hide all but 3 of the existing sites. You should at max have no more than 3 sites shown.
2. Go to `My Site` -> `App Settings` -> `Test feature configuration`.
3. Enable the `OnboardingImprovementsFeatureConfig` feature flag.
4. Log-out and log-in to the app again.
5. Make sure the new `Login Epilogue` screen looks as expected, the `normal` design is shown.

#### Scenario: `expanded`
1. Find an account which has lots of sites and log-in to the app.
2. Go to the `Side Chooser` screen and unhide all sites. You should at least have more that 3 sites shown.
2. Go to `My Site` -> `App Settings` -> `Test feature configuration`.
3. Enable the `OnboardingImprovementsFeatureConfig` feature flag.
4. Log-out and log-in to the app again.
5. Make sure the new `Login Epilogue` screen looks as expected, the `expanded` design is shown.

-----

#### Merge Instructions:
- ~~@osullivanchris to do a design review of both the new `Login Epilogue` screens (`normal` and `expanded`).~~
- ~~Remove the `[Status] Needs Design Review`.~~
- ~~Merge [WordPress-Login-Flow-Android#60](https://github.com/wordpress-mobile/WordPress-Login-Flow-Android/pull/60).~~
- ~~Update `Login` hash from the PR (`60-{sha1}`) to `develop-{sha1}`.~~
- Merge this PR.

-----

#### Notes

This PR is UI related only, the below functionality will be handled in separate PRs:
- Selecting a `Site`, which will either show the `Quick Start Prompt` or navigate to `My Site` directly.
- Clicking the `Create a new site` button, which will start the `WordPress.com` site creation flow.
- `Landscape` UI, which requires a bit more fine tuning before being ready.

-----

## Regression Notes
1. Potential unintended areas of impact

The existing `Login Epilogue` screen might be somehow affected by all the changes in this PR since all 3 screen are sharing all the logic, that is code-wise.

2. What I did to test those areas of impact (or what existing automated tests I relied on)

See `To test` section above.

3. What automated tests I added (or what prevented me from doing so)

N/A

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

ParaskP7 committed:
Refactor: Replace get activity with require activity on login epilogue

ParaskP7 committed:
Refactor: Replace get arguments with require arguments on login epilogue

ParaskP7 committed:
Refactor: Refactor set adapter on login epilogue fragment

This way every parameter ends up being on a separate line, thus making
the 'new SitePickerAdapter(...)' constructor read better.

ParaskP7 committed:
Refactor: Refactor bind header view holder on login epilogue fragment

This way every parameter ends up being on a separate line, thus making
the 'holder.showSitesHeading(...)' call read better.

ParaskP7 committed:
Feature: Create new login epilogue header

This new login epilogue header will be shown when the
'Onboarding Improvements' feature flag gets enabled, else the existing
login epilogue header will be shown.

PS: The design on this new login epilogue header is not final yet. More
refinements are due when the design gets finalized.

ParaskP7 committed:
Feature: Implement new login epilogue header logic

With the new login epilogue header the logic of inflating the layout and
showing/hiding the site headings needs to be altered. This commit is
utilizing the 'Onboarding Improvements' logic to alternate between the
new and existing login epilogue header logic.

ParaskP7 committed:
Feature: Update new login epilogue header design

The design updates is based on final designs as those were provided
by the designer.

ParaskP7 committed:
Refactor: Extract new adapter methods for listener, header and footer

Those newly extracted method will help with readability and making sure
that the logic can be easily updated based on a feature flags being
enabled/disabled, especially for the 'footerHandler()' method.

ParaskP7 committed:
Refactor: Enhance nullability reference within site picker adapter

ParaskP7 committed:
Feature: Hide footer handler when onboarding improvements are enabled

ParaskP7 committed:
Feature: Create new login epilogue screen

This new login epilogue screen will be shown when the
'Onboarding Improvements' feature flag gets enabled, else the existing
login epilogue screen will be shown.

PS: This new login epilogue screen will be shown for 2-3 sites or
less only. For more than 3 sites, another new login epilogue screen will
be shown instead.

ParaskP7 committed:
Feature: Create new expanded login epilogue screen

This new expanded login epilogue screen will be shown when the
'Onboarding Improvements' feature flag gets enabled and when the number
of sites exceeds the expected number, else the existing login epilogue
screen will be shown.

PS: This new expanded login epilogue screen will be shown for more
that 3 sites. For 2-3 sites or less, the non-expanded new login epilogue
screen will be shown instead.

ParaskP7 committed:
Refactor: Explicitly define the divider view field as nullable

This is done since there are actually two dividers, the 'divider' and
the 'item_diver' views, which do not appear on the same screen and as
such when one is present the other one is not, and vice versa.

ParaskP7 committed:
Feature: Hide the item divider on the last item of site picker adapter

This is a new requirement as per the updated design.

ParaskP7 committed:
Refactor: Extract and make get blogs for current view method public

This method will be utilized within the 'LoginEpilogueFragment' to help
identify whether the new login epilogue sceen UI needs to be shown or
the new expanded one instead (for more than 3 sites).

ParaskP7 committed:
Refactor: Explicitly define the bottom shadow view field as nullable

This is done because this 'bottom_shadow' view will not appear on the
same screen every time and it will only appear under sertain condition,
like with the existing UI or the new expanded UI.

ParaskP7 committed:
Refactor: Init site picker adapter during on create

Instead of initializing the 'SitePickerAdapter' during
'setupContent(...)', this is now done during the 'onCreate(...)' phase
of this screen's lifecycle. This is due to the fact that this adapter
and its publicly made 'getBlogsForCurrentView()' method will be required
to be available during 'createMainView(...)' so that the appropriate new
login epiloge screen is being picked up (the normal or expanded one).

ParaskP7 committed:
Feature: Implement new login epilogue screen logic

This commit is utilizing the 'Onboarding Improvements' feature flag to
alternate between the new and existing login epilogue screens.

ParaskP7 committed:
Feature: Implement new expanded login epilogue screen logic

This commit is utilizing the 'EXPANDED_UI_THRESHOLD' logic to
alternate between the expanded and normal new login epilogue screens.

ParaskP7 committed:
Refactor: Removed unused returned value from init adapter

osullivanchris reviewed (changes requested):
Thanks for sharing @ParaskP7 , its looking great already. Here are a few things I noticed:

- the dividing line is going straight through the 'or' text, rather than breaking either side
- in landscape mode, with only one site, the 'create site' button is not sticky and is off screen for me. I think it should always be sticky in landscape mode for both 1-3 sites and more site versions.
- the buttons are not wired up yet. I couldn't tap a site. 'Create new site' took me to My Site. Assume this is just not done yet.
- Seeing the 'Choose a site to open' text in serif, I'm not sure we've got the balance right with type sizes yet. Let's discuss options together next week.

ashiagr commented on WordPress/src/main/res/layout/login_epilogue_screen_new.xml:
```diff
@@ -0,0 +1,72 @@
+        app:layout_constraintStart_toStartOf="parent"
+        app:layout_constraintTop_toTopOf="parent" />
+
+    <FrameLayout
+        android:id="@+id/orLayout"
+        android:layout_width="match_parent"
+        android:layout_height="wrap_content"
+        android:layout_marginTop="@dimen/margin_extra_medium_large"
+        android:paddingBottom="@dimen/margin_medium"
+        android:paddingTop="@dimen/margin_medium"
+        app:layout_constraintBottom_toTopOf="@+id/bottom_button"
+        app:layout_constraintEnd_toEndOf="parent"
+        app:layout_constraintStart_toStartOf="parent"
+        app:layout_constraintTop_toBottomOf="@+id/recycler_view">
+
+        <TextView
+            style="@style/Widget.LoginFlow.TextView.DividerText"
+            android:layout_width="wrap_content"
+            android:layout_height="wrap_content"
+            android:layout_gravity="center_horizontal"
+            android:paddingEnd="@dimen/margin_small"
+            android:paddingStart="@dimen/margin_small"
+            android:text="@string/login_or" />
+
+        <View
+            android:layout_width="match_parent"
+            android:layout_height="1dp"
+            android:layout_gravity="center_vertical"
+            android:alpha="0.12"
+            android:background="?attr/colorOnSurface" />
+
+    </FrameLayout>
```
Suggestion 💡: Since this `orLayout` is now used at three places:

`login_email_screen.xml`
`login_epilogue_screen_new.xml`
`login_epilogue_screen_new_expanded.xml`

how about it is extracted to a separate layout and included in all three places? This way, we wouldn't have to worry about the dividing line going through the "or" text as noticed by Chris in the future.

ParaskP7 replied:
👋 @ashiagr !

Such a nice suggestion, totally, I agree 💯 and will do that asap!

ParaskP7 replied:
👋 @ashiagr !

I have created this [WordPress-Login-Flow-Android#60](https://github.com/wordpress-mobile/WordPress-Login-Flow-Android/pull/60) PR which extract the `login-or-layout` as discussed. Let me know your thoughts. Then, I'll incorporate this change within this PR.

PS: @osullivanchris , this [WordPress-Login-Flow-Android#60](https://github.com/wordpress-mobile/WordPress-Login-Flow-Android/pull/60) PR above slightly changes the UI on the `login-email-screen` for just a `2dp` bottom diff, between the `Or Layout` and the `Google Sign-In` button. This change is barely noticeable but I chose to make it to make this `Or Layout` consistent across all screens afterwards. Are you okay with this change? PS: When this change gets incorporated in this PR I can ping you again so that you can double check the `login-email-screen` and see if you don't like something, maybe those extra `2dp` will create an issue of some kind, design wise.

osullivanchris replied:
@ParaskP7 that sounds fine. Good to be consistent and stick to the 8dp grid. But I can take a look at a build/screenshot of both screens when its done just to be sure and give it a final 👍

ParaskP7 replied:
👋 @osullivanchris !

Yes, exactly, true. I'll then ping you when we get ready with the final APK and I'll have the updated screenshots.

PS: Bear in mind that this might take sometime since we are having some issues with the `Login` library but `Platform9` is working on fixing them as we speak.

ParaskP7 replied:
👋 @ashiagr !

I have updated the PR with the extracted `login-or-layout` that is not being reused, could you please do another review on that: fbc7089c325b5aefee0c579dd102d1c1665655f0 🙏 

👋 @osullivanchris !

I have update the PR description screenshots to the how the `Or Layout` is looking now. Can you also please run through the app and verify that everything works as expected, including looking at the `login-email-screen` with the `Google Sign-In` button and its `Or Layout` and verify that it looks good as well? 🙏

ashiagr replied:
👋 @ParaskP7,

I've reviewed fbc7089, also tested all screens including this layout, all LGTM 👍. 
I've approved the PR, feel free to merge once Chris approves it.

ParaskP7 replied:
Thank you @ashiagr ! 🙏

osullivanchris replied:
> I have update the PR description screenshots to the how the Or Layout is looking now. Can you also please run through the app and verify that everything works as expected, including looking at the login-email-screen with the Google Sign-In button and its Or Layout and verify that it looks good as well? 🙏

@ParaskP7 you may have finally found something that is beyond my eye to spot 🤓  I can't tell, so its totally fine to go with this change if it improves the reusability of this pattern.

ParaskP7 replied:
👋 @osullivanchris !

> @ParaskP7 you may have finally found something that is beyond my eye to spot 🤓 I can't tell, so its totally fine to go with this change if it improves the reusability of this pattern.

Nice, thanks so much for double checking this Chris, and for the 👍 ! We are now good to 🚀

ashiagr commented on WordPress/src/main/java/org/wordpress/android/ui/accounts/login/LoginEpilogueFragment.java:
```diff
@@ -92,17 +98,30 @@ protected void setupLabel(@NonNull TextView label) {
+
+    @LayoutRes
+    private int loginEpilogueScreenResource() {
+        if (mOnboardingImprovementsFeatureConfig.isEnabled()) {
+            if (mAdapter.getBlogsForCurrentView().size() <= EXPANDED_UI_THRESHOLD) {
+                return R.layout.login_epilogue_screen_new;
+            } else {
+                return R.layout.login_epilogue_screen_new_expanded;
+            }
+        } else {
+            return R.layout.login_epilogue_screen;
+        }
```
Warning ⚠️:  IIRC we discussed not making any changes to the epilogue screen for the Jetpack app or at least not show the create site option. I'd suggest either we update the condition for the appropriate layout display in this PR or add a note in the task to not forget about it.

ParaskP7 replied:
👋 @ashiagr !

Great observation, I added a additional todo task on the parent issue [here](https://github.com/wordpress-mobile/WordPress-Android/issues/14990), which will deal with everything `Jetpack App` related. I think it is better for us to dedicate some additional time, add more thought and testing to when it comes to the Jetpack app. This way, it will also make this PR less complex and have it focusing on the WordPress app alone, at least until the full solution is complete. Wdyt?

ashiagr replied:
Works for me 🙂, thanks for adding the additional todo task 🙏.

ashiagr reviewed (commented):
Great job @ParaskP7! Changes are coming along really nicely  🙌. Also, appreciate the in-between refactoring you did to make the code more readable and robust. 

I only left one suggestion and a warning for you to consider. Hope it helps.

ParaskP7 commented:
👋 @osullivanchris !

Thanks for much for this detailed review! 🙏 

> - the dividing line is going straight through the 'or' text, rather than breaking either side

You are totally right, really good eye, I totally missed that. It is very strange that this is happening since I copy-pasted the component from the `Google` sign-in form as it was specified there, thus I never even bother to notice it afterwards. Hmmm... I wonder... I'll update it asap and let you know! 👍 

> - in landscape mode, with only one site, the 'create site' button is not sticky and is off screen for me. I think it should always be sticky in landscape mode for both 1-3 sites and more site versions.
> - the buttons are not wired up yet. I couldn't tap a site. 'Create new site' took me to My Site. Assume this is just not done yet.

I have left a note on my description on all that. `Selecting a Site`, `Clicking the Create a new site` button and the `Landscape UI` will be done in individual PR, separate to this one, you can also check the list of todos in the parent issue, [here](https://github.com/wordpress-mobile/WordPress-Android/issues/14990). Also, I just added the `Jetpack App` todo item, which will be done in a separate PR as well. Divide and conquer! 😄 

> - Seeing the 'Choose a site to open' text in serif, I'm not sure we've got the balance right with type sizes yet. Let's discuss options together next week.

Definitely, let's discuss our options together. In this instance I used the `@style/Widget.LoginFlow.Button.Secondary` style that is already the style for the secondary (`white`) button during the login flow, as it is already defined by the library. I didn't want to make any changes to that before you can see the result and suggest updates to the button. 👍

osullivanchris commented:
Thanks for the reply @ParaskP7 

> You are totally right, really good eye, I totally missed that. It is very strange that this is happening since I copy-pasted the component from the Google sign-in form as it was specified there, thus I never even bother to notice it afterwards. Hmmm... I wonder... I'll update it asap and let you know! 👍

strange! Thanks for taking a look into it!

> I have left a note on my description on all that. Selecting a Site, Clicking the Create a new site button and the Landscape UI will be done in individual PR, separate to this one, you can also check the list of todos in the parent issue, here. Also, I just added the Jetpack App todo item, which will be done in a separate PR as well. Divide and conquer! 😄

Nice! Sorry I missed that. I'll have a look at the parent task!

> Definitely, let's discuss our options together. In this instance I used the @style/Widget.LoginFlow.Button.Secondary style that is already the style for the secondary (white) button during the login flow, as it is already defined by the library. I didn't want to make any changes to that before you can see the result and suggest updates to the button. 👍

Is that the button you're referring to? I was talking about the subheading. We can discuss on our call in a bit 😄

ParaskP7 commented:
Thank YOU @osullivanchris !

> strange! Thanks for taking a look into it!

True, it was just our IDE messing with reordering XML elements automatically, which is causing that, and this is because of the predefined setup we are all using. We also talked with @ashiagr to extract this login in a separate layout and make it reusable. This will avoid further such problems in the future and make it easier for anyone else to reuse this component. 👍 

> Nice! Sorry I missed that. I'll have a look at the parent task!

The description was long, of course! 🌟 

> Is that the button you're referring to? I was talking about the subheading. We can discuss on our call in a bit 😄

Oh, you are right, I misread, this is not about the `button`, but rather the `subheading`. In any ways, I am open for any suggestions you might have for me! 👍

ParaskP7 committed:
Build: Update login hash to PR#60

This 'Login' hash updates 'Login' library to that 'PR' version where
the 'login_or_layout' layout has been extracted and can be now reused.

ParaskP7 committed:
Refactor: Replace hardcoded or layout with extracted login or layout

ParaskP7 committed:
Fix: Remove unexpected character within login epilogue screen new layout

ParaskP7 committed:
Fix: Update text appearance to subtitle 1 for choose a site to open

After talking with the designer it was decided that the best choice is
to go with the 'textAppearanceSubtitle1' for now.

ParaskP7 commented:
👋 @osullivanchris !

> Oh, you are right, I misread, this is not about the button, but rather the subheading. In any ways, I am open for any suggestions you might have for me! 👍

As discussed, I have just pushed a fix which updates the `textAppearanceHeadline6` text appearance to `textAppearanceSubtitle1`. I have also updated the description with the new screenshots. Please do let me know if everything looks ok now and thank you a lot for all your support. 🙏

osullivanchris commented:
Looks good @ParaskP7 and thanks for the chat! I'm updating zeplin to reflect the same change 😄

ParaskP7 committed:
Build: Update login hash to the latest develop version

This 'Login' hash updates 'Login' library to that 'develop' version
where the 'login_or_layout' layout has been extracted and can be now
reused.
